code_annotations/find_annotations.py

```diff
@@ -0,0 +1,254 @@
+    """
+    Callback for when we fail to load an extension, dumps the error and
+    raises an exception. By default these errors just fail silently.
+
+    :param args:
+    :param kwargs:
+    :raises: ConfigurationException
+    """
+    print(args)
+    print(kwargs)
+    raise Exception('Failed to load a plugin, aborting.')
+
+
+def search_extension(ext, file_handle, file_extensions_map, filename_extension):
+    """
+    Executes an extension search on the given file, only if it is
+    configured for this extension
+
+    :param ext:
+    :param file_handle:
+    :param file_extensions_map:
+    :param filename_extension:
+    :return: Tuple of (extension name, list of found annotation dicts)
+    """
+    # Reset the read handle to the beginning of the file in case another
+    # extension already moved it
+    file_handle.seek(0)
+
+    # Only search this file if we are configured for its extension
+    if filename_extension in file_extensions_map[ext.name]:
+        ext_results = ext.obj.search(file_handle)
+
+        if ext_results:
+            return ext.name, ext_results
+
+    return ext.name, None
+
+
+def format_file_results(source_path, all_results, results):
+    """
+    Adds all extension search results for a single file into the current
+    search results dict, indexed by filename
+
+    :param source_path: Top level search path
+    :param all_results: Aggregated results to add the results to
+    :param results: Results of search() on a single file
+    :return: None, modifies all_results
+    """
+    # _ here is the extension name, as required by Stevedore map(). Each
+    # annotation already has the extension name so we can ignore it here
+    for _, annotations in results:
+        if annotations is None:
+            continue
+
+        # TODO: The file_path should be the same for all of these results
+        # so we should be able to optimize getting file_path and making
+        # sure it exists in the dict to do this less often.
+
+        # source_path is unnecessary here, we only want the file name
+        # relative to it.
+        file_path = annotations[0]['filename'].replace(source_path, '')
+
+        if file_path not in all_results:
+            all_results[file_path] = []
+
+        # TODO: add support for multiple extensions in the 'found_by' key
+        # and de-dupe results
+        all_results[file_path].extend(annotations)
+
+
+def configure(config, source_path, report_path, verbose):
+    """
+    Reads the configuration file, and handles command line overrides
+
+    TODO: Add include / exclude directories
+
+    :param config: Location of the configuration file
+    :param source_path: Path to the code to be searched
+    :param report_path: Directory where the report will be generated
+    :param verbose: Bool indicating verbosity level
+    :return: Configuration dict, updated with overrides
+    """
+
+    # TODO: Check for configuration file existence for prettier error
```

Comment:
My preference is to utilize click's `click.Path(exists=True, dir_okay=False)` to do existence checks for us, such as in https://github.com/edx/code-annotations/pull/4/files#diff-6c3716bd15107be0162290a140490c82R25